### PR TITLE
Populate shoot.status.advertisedAddresses early in the reconciliation flow

### PR DESCRIPTION
**Pull request by vpnachev**
**How to categorize this PR?**

/kind enhancement

**What this PR does / why we need it**:
Populate `shoot.status.advertisedAddresses` early in the reconciliation flow instead of at the end or on failure.

_(edit by @voelzmo: inserted this from [this comment below](https://github.com/gardener/gardener/pull/4963#issuecomment-964092624) for future reference)_
We have a tool deploying gardener landscapes. This tool is creating shoot clusters asynchronously (it does not wait the shoot creation to be completed) and exports OIDC kubeconfigs for them. In the kubeconfig we need to set the address of the API server. As there are dns-less environments or some dns defaulting is applied, the tool does not know about the address in advance, instead it has to read it from the Gardener API. To improve and speed-up this flow, I propose this PR so that advertised addresses are set right after the kube-apiserver service is ready instead of the current behavior - setting them after the exit of the reconciliation flow.

/invite @dimityrmirchev 

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```other operator
The `shoot.status.advertisedAddresses` are now set as early as possible in the shoot reconciliation flow.
```

### CODE REVIEW

**Comment by vpnachev** [comment]
/hold

As it is not working as expected with `unmanaged` address.

**Comment by voelzmo** [comment]
Hey @vpnachev can you help me understand the context on this PR a bit better? Is this related to an issue we saw or some other kind of functionality needs this change?

**Review by stoyanr — APPROVED** [review]
/lgtm 
The change makes sense in terms of consistency (a field in the status is updated as early as possible instead of waiting until the flow finishes with either success or error). Similarly to @voelzmo I am also wondering if there is impact beyond that?

**Comment by vpnachev** [comment]
We have a tool deploying gardener landscapes. This tool is creating shoot clusters asynchronously (it does not wait the shoot creation to be completed) and exports OIDC kubeconfigs for them. In the kubeconfig we need to set the address of the API server. As there are dns-less environments or some dns defaulting is applied, the tool does not know about the address in advance, instead it has to read them from the Gardener API. To improve and speed-up this flow, I propose this PR so that advertised addresses are set right after the kube-apiserver service is ready instead of the current behavior - setting them after the exit of the reconciliation flow.

**Comment by stoyanr** [comment]
@vpnachev Thanks! 
@voelzmo Any objections to merge this? I think the explanation provided is more than adequate.

**Review by voelzmo — APPROVED** [review]
/lgtm

Thanks for the explanations!

**Comment by dimityrmirchev** [comment]
/lgtm

Nice!

**Review by dimityrmirchev — APPROVED** [review]

**Review by timebertt — APPROVED** [review]
Thanks for the explanation @vpnachev, that was helpful :)
/lgtm

**Pull request merged**
